File: that.py
import sys
from collections import deque

class thatParser(object):
    def __init__(self):
        self.BEGIN = ('this!', 'this?')
        self.END = ('this?', 'this!')
        self.primitives = {('this.', 'this.'): 'inc',
                           ('this!', 'this!'): 'dec',
                           ('this.', 'this?'): 'right',
                           ('this?', 'this.'): 'left',
                           ('this!', 'this.'): 'write',
                           ('this.', 'this!'): 'read'}
                           
    def parse(self, input_text):
        items = input_text.lower().split()
        for i in range(0, len(items), 2):
            x = (items[i], items[i+1])
            if x in self.primitives or x in (self.BEGIN, self.END):
                yield x
                
class Interpreter(object):
    MAX_NESTED_LOOPS = 1000
    def __init__(self):
        self.that_parser = thatParser()
        self.set_parser(self.that_parser)

    def reset(self):
        self.cells = deque([0])
        self.index = 0
        self.input_buffer = []
        self.output_buffer = []
        self.open_loops = 0
        self.loop = []

    def inc(self):
        self.cells[self.index] += 1

    def dec(self):
        self.cells[self.index] -= 1

    def right(self):
        self.index += 1
        if self.index >= len(self.cells):
            self.cells.append(0)

    def left(self):
        if self.index == 0:
            self.cells.appendleft(0)
        else:
            self.index -= 1

    def write(self):
        self.output_buffer.append(self.cells[self.index])
            
    def read(self):
        try:
            self.cells[self.index] = int(raw_input("Your input: "))
        except (TypeError, ValueError):
            print "oops"

    def as_ascii(self):
        return "".join([chr(c) for c in self.output_buffer])

    def set_parser(self, parser):
        self.parser = parser
        self.reset()

    def interpret_raw_text(self, text):
        self.input_buffer.extend(self.parser.parse(text))
        try:
            self.interpret_items(self.input_buffer)
            self.input_buffer = []
        except IndexError:
            print "oops"

    def interpret_items(self, items):
        for item in items:
            if self.open_loops:
                self.interpret_inside_loop(item)
            else:
                self.interpret_directly(item)

    def interpret_inside_loop(self, item):
        if item == self.parser.END:
            self.open_loops -= 1
            if self.open_loops == 0:
                while self.cells[self.index]:
                    self.interpret_items(self.loop)
                return
        elif item == self.parser.BEGIN:
                self.open_loops += 1
        self.loop.append(item)

    def interpret_directly(self, item):
        if item == self.parser.END:
            raise ValueError("End without begin")
        elif item == self.parser.BEGIN:
            self.open_loops = 1
            self.loop = []
        elif item in self.parser.primitives:
            method = self.parser.primitives[item]
            getattr(self, method)()
        else:
            print "Unknown token '%s' - ignored" % (item, )

    def interpret_file(self, fname):
        file = open(fname, 'r')
        self.interpret_raw_text(file.read())

    def __repr__(self):
        rep = "\n".join(["ASCII output\t: %s"])

        return rep % (self.as_ascii())

def print_usage():
    print "\nUsage:"
    print "Interpret That file: python that.py <FILENAME>\n"


if __name__ == '__main__':
    if len(sys.argv) == 2:
        that = Interpreter()
        that.interpret_file(sys.argv[1])
        print that
    else:
        print_usage()

File: that.js
var begin = "thisthisthis";
var end = "   ";
var inc = "thisthis ";
var dec = " thisthis";
var right = "  this";
var left = "this  ";
var write = "this this";
var read = " this ";

this.cells = [0];
this.index = 0;
this.input_buffer = [];
this.output_buffer = [];
this.open_loops = 0;
this.loop = [];

function That(){
	
	this.brainFuck2That = function(){
		console.log("brainFuck2That");
	};
	this.that2BrainFuck = function(){
		console.log("that2BrainFuck");
	};
	this.that2Ascii = function(){
		var input = document.getElementById("That").value;
		this.resetParser();
		while(input.length > 0){
			if(input.indexOf(begin) === 0){
				console.log("found begin!")
				input = input.substring(12);
				this.begin();
			}
			else if(input.indexOf(end) === 0){
				console.log("found end!")
				input = input.substring(3);
				this.end();
			}
			else if(input.indexOf(inc) === 0){
				console.log("found inc!")
				input = input.substring(9);
				this.inc();
			}
			else if(input.indexOf(dec) === 0){
				console.log("found dec!")
				input = input.substring(9);
				this.dec();
			}
			else if(input.indexOf(right) === 0){
				console.log("found right!")
				input = input.substring(6);
				this.right();
			}
			else if(input.indexOf(left) === 0){
				console.log("found left!")
				input = input.substring(6);
				this.left();
			}
			else if(input.indexOf(write) === 0){
				console.log("found write!")
				input = input.substring(9);
				this.write();
			}
			else if(input.indexOf(read) === 0){
				console.log("found read!")
				input = input.substring(6);
				this.read();
			}
			else{
				input = "";
				console.log("Invalid Syntax");
			}
		}
		var output = this.asAscii();
		document.getElementById("Ascii").value = output;
	};
	this.asAscii = function(){
		console.log(this.output_buffer);
		var asciiArr = [];
		for(var i = 0; i < this.output_buffer.length; i++){
			asciiArr[i] = String.fromCharCode(this.output_buffer[i]);
		}
		return asciiArr.join("");
	};
	
	this.ascii2That = function(){
		var input = document.getElementById("Ascii").value;
		this.output = "";
		for(var i = 0; i < input.length; i++){
			for(var j = 0; j < input.charCodeAt(i); j++){
				this.output = this.output.concat(inc);
			}
			this.output = this.output.concat(write);
			this.output = this.output.concat(right);
		}
		document.getElementById("That").value = this.output;
	};	
	
	this.resetParser = function(){
        this.cells = [0];
		this.index = 0;
		this.input_buffer = [];
		this.output_buffer = [];
		this.open_loops = 0;
		this.loop = [];
	};
	
	this.inc = function(){
		this.cells[this.index] += 1;
	};
	
	this.dec = function(){
		this.cells[this.index] -= 1;
	};
	
	this.right = function(){
        this.index += 1;
		if(this.index === this.cells.length){
			this.cells.push(0);
		}
	};
	
	this.left = function(){
		if(this.index === 0){
			this.cells.unshift(0);
		}
		else{
			this.index -= 1;
		}
	};
	
	this.write = function(){
		this.output_buffer.push(this.cells[this.index]);
	};
	
	this.read = function(){
		console.log("Reading not implemented");
	};
	
	this.begin = function(){
		console.log("Loops not implemented");
	};
	
	this.end = function(){
		console.log("Loops not implemented");
	};
}
